File: Leetcode 3 -- Longest Substring Without Repeating Characters.py
# Leetcode 3: Longest Substring without Repeating Characters
# https://leetcode.com/problems/longest-substring-without-repeating-characters/
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        #Method 1: sliding window method
        left = answer = 0
        seen = set()

        for right in range(len(s)):
            # checking for duplicates in substring
            while s[right] in seen:
                seen.remove(s[left])
                left += 1
            seen.add(s[right])
            #updating the answer
            answer = max(answer, right - left + 1)
        
        return answer
        # time complexity: O(N)
        # space complexity: O(N)

        

File: Leetcode 3 -- Longest Substring Without Repeating Characters.java
//Leetcode 3: Longest substring without repeating characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

//Method 1: 2 pointers
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set <Character> seen = new HashSet <Character> ();
        int left = 0;
        int answer = 0;
        
        for (int right = 0;right < s.length(); right++){
            while(seen.contains(s.charAt(right))){
                seen.remove(s.charAt(left));
                left++;
            }
            seen.add(s.charAt(right));
            answer = Math.max(answer, right - left + 1);
        }

        return answer;

    }
}

//time complexity: O(N)
//space complexity: O(1) since there's only 26 characters or 128 if using ASCII
